pkg/kubelet/container/runtime_cache_test.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package container
+
+import (
+	"reflect"
+	"testing"
+	"time"
+)
+
+func TestGetPods(t *testing.T) {
+	runtime := &FakeRuntime{}
+	expected := []*Pod{{ID: "1111"}, {ID: "2222"}, {ID: "3333"}}
+	runtime.Podlist = expected
+	cache, err := NewRuntimeCache(runtime)
+	if err != nil {
+		t.Errorf("unexpected error %v", err)
+	}
+	actual, err := cache.GetPods()
+	if err != nil {
+		t.Errorf("unexpected error %v", err)
+	}
+	if !reflect.DeepEqual(expected, actual) {
+		t.Errorf("expected %#v, got %#v", expected, actual)
+	}
+}
+
+func TestForceUpdateIfOlder(t *testing.T) {
+	runtime := &FakeRuntime{}
+	c, err := NewRuntimeCache(runtime)
+	if err != nil {
+		t.Errorf("unexpected error %v", err)
+	}
+	cache := c.(*runtimeCache)
+
+	// Cache old pods.
+	oldpods := []*Pod{{ID: "1111"}}
+	runtime.Podlist = oldpods
+	cache.updateCacheWithLock()
+
+	// Update the runtime to new pods.
+	newpods := []*Pod{{ID: "1111"}, {ID: "2222"}, {ID: "3333"}}
+	runtime.Podlist = newpods
+
+	// An older timestamp should not force an update.
+	cache.ForceUpdateIfOlder(time.Now().Add(-20 * time.Minute))
+	actual := cache.getCachedPods()
+	if !reflect.DeepEqual(oldpods, actual) {
+		t.Errorf("expected %#v, got %#v", oldpods, actual)
+	}
+
+	// A newer timestamp should force an update.
+	cache.ForceUpdateIfOlder(time.Now().Add(20 * time.Second))
+	actual = cache.getCachedPods()
+	if !reflect.DeepEqual(newpods, actual) {
+		t.Errorf("expected %#v, got %#v", newpods, actual)
+	}
+}
+
+func TestUpdatePodsOnlyIfNewer(t *testing.T) {
+	runtime := &FakeRuntime{}
+	c, err := NewRuntimeCache(runtime)
+	if err != nil {
+		t.Errorf("unexpected error %v", err)
+	}
+	cache := c.(*runtimeCache)
+
+	// Cache new pods with a future timestamp.
+	newpods := []*Pod{{ID: "1111"}, {ID: "2222"}, {ID: "3333"}}
+	cache.Lock()
+	cache.pods = newpods
+	cache.cacheTime = time.Now().Add(20 * time.Minute)
```

Comment:
Ohh... I wasn't actually talking about the unit tests. If I add util.Clock in the runtimeCache itself, I'd expect all the public methods to provide relative time (e.g. time.Duration), but not the absolute time. Since we expose methods that accepts caller's time, it's strange to keep an internal clock that we don't completely rely on.